### PR TITLE
refactor: New test setup for stream handler tests

### PR DESCRIPTION
**Pull request by stiegerc**
This provides a new framework for `StreamHandler` tests and refactors the tests to use it.

`StreamHandler` tests all work with the same pattern:
- A setup stage where loopback/outgoing streams are populated with messages and signals. Same for incoming stream slices.
- A stage where the expected state changes are produced, i.e. we expect some messages to be inducted, others are rerouted or gc'ed and so forth.
- A call to a function of the `StreamHandler` using the state produced in stage 1.
- A comparison of the resulting state after this call to the expected state produced in stage 2.
- A check on the metrics. What was inducted? Which critical errors were raised and which were not raised and so on.

The previous setup was inconsistent with the `CallbackId's` since they were not registered properly. The new setup takes a number of 'configs' to produce messages in streams and stream slices together with a canister state, such that all `CallbackId's` are properly registered and all the messages can be inducted at will.

Once a state prepared for the test has been produced in this way, an expected state can be produced by cloning the initial state, generating new streams by directly cloning messages from streams in the initial state or generating reject responses (in stage 2).

Some other refactoring was added to remove boiler plate and make the tests more readable and to the point. On average tests are 20% to 30% shorter than before.

### CODE REVIEW

**Comment by stiegerc** [comment]
Some details to add:
- Before some tests checked only that a specific error counter was incremented. But it seems equally important to check that counters that you don't expect were not incremented. I introduced a struct where you can specify some counters and then extend with the default (i.e. 0). This way all counters are always checked without adding boilerplate everywhere.
- Before sometimes messages were just generated using `test_response()` or `test_request()` again for the expected streams, but this makes it harder to follow what exactly is rerouted and whatnot. Not messages are explicitly cloned from a stream at a certain index. In theory this could even shroud an ordering violation in the requests. Since all messages are now unique due to the correctly assigned `CallbackId's` this couldn't happen anymore.
- Before it created the state in isolation and only then the streams. So something like `let expected_state = state.clone()` cloned an empty state or with 1 canister installed. Since it now generates the entire test setup at once and returns it, this line clones the streams as well. Later on we replace the initial streams with the expected streams we generated using `expected_state..with_streams(..)`. This overwrites all streams. I wasn't sure whether this might be confusing or not. But it seems to me that the way I have it now is the clearest way (also the tiny hit in performance is irrelevant).

**Comment by stiegerc** [comment]
Note that I use 'pretty_assertions' here.

`assert_eq!(expected_state, inducted_state)` produces unreadable walls of text making debugging quite painful. At some point I figured I could just ask chatGPT what the difference was on the output and then it suggested this crate. Great suggestion!

File: rs/messaging/src/routing/stream_handler/tests.rs [hunk 1]
```diff
@@ -689,1112 +539,1230 @@ fn induct_loopback_stream_with_memory_limit_impl(
...
+    let inducted_state = stream_handler.process_stream_slices(state, slices);
+
+    assert_eq!(inducted_state, expected_state);
 }
 
 #[test]
-fn reroute_rejected_response_success() {
-    with_test_replica_logger(|log| {
-        // Empty initial state.
-        let (stream_handler, mut initial_state, _) = new_fixture(&log);
```
Comment (stiegerc) [inline rs/messaging/src/routing/stream_handler/tests.rs:1699]: This test is redundant since what it tests is already included in `garbage_collect_local_state_with_reject_signals_for_response_success()`. It is the only test that tests one of the 'internal functions' of the stream handler directly. It therefore also doesn't quite fit with my test framwork very well.

File: rs/messaging/src/routing/stream_handler/tests.rs [hunk 2]
```diff
@@ -1883,439 +1857,228 @@ fn check_stream_handler_generated_reject_response_canister_migrating() {
...
-fn induct_stream_slices_response_to_missing_canister() {
-    with_test_replica_logger(|log| {
-        let (stream_handler, mut initial_state, metrics_registry) = new_fixture(&log);
-
-        let mut expected_state = initial_state.clone();
-
-        // Initial state with no canisters and one stream.
-        let outgoing_stream = generate_outgoing_stream(StreamConfig {
-            messages_begin: 21,
-            message_count: 0,
```
Comment (stiegerc) [inline rs/messaging/src/routing/stream_handler/tests.rs:2074]: This test is redundant since it's functionality is completely encompassed by the more complex test `induct_stream_slices_partial_success()`. The only thing this test was missing was the check on the critical error counter, which should have been there anyway. Considering there are even more complicated tests in this file, I see no reason to check this in isolation.

File: rs/messaging/src/routing/stream_handler/tests.rs [hunk 3]
```diff
@@ -1883,439 +1857,228 @@ fn check_stream_handler_generated_reject_response_canister_migrating() {
...
-#[test]
-fn induct_stream_slices_sender_subnet_mismatch() {
-    with_test_replica_logger(|log| {
-        let (stream_handler, mut initial_state, metrics_registry) = new_fixture(&log);
-
-        // Initial state with one canister...
-        let initial_canister_state = new_canister_state(
-            *LOCAL_CANISTER,
-            user_test_id(24).get(),
-            *INITIAL_CYCLES,
```
Comment (stiegerc) [inline rs/messaging/src/routing/stream_handler/tests.rs:2152]: Same for this test. Also included in `induct_stream_slices_partial_success()`.

File: rs/messaging/src/routing/stream_handler/tests.rs [hunk 4]
```diff
@@ -83,520 +90,390 @@ fn oversized_reject_message_is_truncated() {
...
 )]
-fn induct_loopback_stream_with_signals_panics() {
-    with_test_replica_logger(|log| {
-        let (stream_handler, mut state, _) = new_fixture(&log);
-
-        // A loopback stream containing 2 messages and 1 signal.
-        let loopback_stream = generate_loopback_stream(StreamConfig {
-            messages_begin: 21,
-            message_count: 2,
+fn induct_loopback_stream_with_signals_end_before_messages_end_panics() {
```
Comment (alin-at-dfinity) [inline rs/messaging/src/routing/stream_handler/tests.rs:99]: Nit: I find the earlier "loopback stream with signals panics" description a lot easier to grasp than "`signals_end < end` (this must be equal...".

Feel free to add a doc comment explaining why we expect no signals in the loopback stream, but keep it at a high level (i.e. "no signals") instead of putting it in terms of the implementation ("this fields < that field because ...").

File: rs/messaging/src/routing/stream_handler/tests.rs [hunk 5]
```diff
@@ -83,520 +90,390 @@ fn oversized_reject_message_is_truncated() {
...
+            ..StreamConfigX::default()
+        }]);
 
-        stream_handler.induct_loopback_stream(state, &mut 1_000_000_000);
-    });
+    stream_handler.induct_loopback_stream(state, &mut (i64::MAX / 2));
 }
 
+/// Tests that inducting a loopback stream where `signals_end` < `begin` panics (this must be equal
+/// for the loopback stream, because it is its own reverse stream).
```
Comment (alin-at-dfinity) [inline rs/messaging/src/routing/stream_handler/tests.rs:112]:
```suggestion
/// Tests that inducting a loopback stream where `signals_end` < `begin` panics: messages are always
/// garbage collected up to `signals_end` and `signals_end` is monotonically increasing.
```

File: rs/messaging/src/routing/stream_handler/tests.rs [hunk 6]
```diff
@@ -83,520 +90,390 @@ fn oversized_reject_message_is_truncated() {
...
-
-        // A loopback stream containing 2 messages and 1 signal.
-        let loopback_stream = generate_loopback_stream(StreamConfig {
-            messages_begin: 21,
-            message_count: 2,
+fn induct_loopback_stream_with_signals_end_before_messages_end_panics() {
+    let (stream_handler, state, _) =
+        with_local_test_setup(btreemap![LOCAL_SUBNET => StreamConfigX {
+            begin: 21,
+            messages: vec![*LOOPBACK_REQUEST_CONFIG; 2],
```
Comment (alin-at-dfinity) [inline rs/messaging/src/routing/stream_handler/tests.rs:103]: Nit: I'd rather spell out `[*LOOPBACK_REQUEST_CONFIG, *LOOPBACK_REQUEST_CONFIG]`, it takes a lot less effort to eyeball than `[*LOOPBACK_REQUEST_CONFIG; 2]`.

Also, looking at it in this context, whether the struct is called `MessageConfigX` or not, it makes a lot more sense for the constant / static variable to be called `LOOPBACK_REQUEST` than `LOOPBACK_REQUEST_CONFIG`. (Or you can just replace it with something like `request(*LOCAL_CANISTER, *LOCAL_CANISTER)`, as suggested elsewhere. Up to you.)

File: rs/messaging/src/routing/stream_handler/tests.rs [hunk 7]
```diff
@@ -83,520 +90,390 @@ fn oversized_reject_message_is_truncated() {
...
             signals_end: 20,
-            reject_signals: None,
-        });
-        state.with_streams(btreemap![LOCAL_SUBNET => loopback_stream]);
+            ..StreamConfigX::default()
+        }]);
 
-        stream_handler.induct_loopback_stream(state, &mut 1_000_000_000);
-    });
+    stream_handler.induct_loopback_stream(state, &mut (i64::MAX / 2));
```
Comment (alin-at-dfinity) [inline rs/messaging/src/routing/stream_handler/tests.rs:126]: Only just noticed this: `with_test_replica_logger()` expects all of the work to happen within the closure; calling `induct_loopback_stream()` (or whatever else, in other tests) outside `with_test_replica_logger()` will result in any log messages produced by the test code never being printed.

I.e. you need to refactor your `with_*_test_setup_*()` functions to also take the test body as a closure and execute it within `with_test_replica_logger()`. Something like:

```rust
with_local_test_setup(
    btreemap![LOCAL_SUBNET => StreamConfigX {
        ...
    }],
    |stream_handler, state, _slies, _log| {
        stream_handler.induct_loopback_stream(state, &mut (i64::MAX / 2));
    }
)
```

Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler/tests.rs:126]: This is what I wanted to do at first but then didn't because the formatter indents it more than I was hoping for.  After playing around with it a bit, I am fairly certain that you can't prevent this unless perhaps with a macro. But I guess this extra indentation isn't so bad.

File: rs/messaging/src/routing/stream_handler/tests.rs (hunk 4 above)
Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler/tests.rs]: Changed.

File: rs/messaging/src/routing/stream_handler/tests.rs (hunk 5 above)
Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler/tests.rs]: Changed.

File: rs/messaging/src/routing/stream_handler/tests.rs (hunk 6 above)
Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler/tests.rs]: I'm not using the `vec![a; 2]` syntax anymore though I don't see what the problem is TBH.

`MessageConfigX` is now called `MessageBuilder` since it's basically a wrapper around `RequestBuilder`, `ResponseBuilder`.

Is is then later on used in `StreamConfig` like so:
```
StreamConfig {
   messages: vec![
      BuildRequest(*LOCAL_CANISTER, *LOCAL_CANISTER),
      BuildResponse(*LOCAL_CANISTER, *OTHER_LOCAL_CANISTER),
   ],
   ..StreamConfig::default()
}
```
Which is supposed to mean: 'when you make a stream for me please build a request from here to there and build a response from here to there`...

I also ditched the constants for `LOOPBACK_REQUEST` and so on.

File: rs/messaging/src/routing/stream_handler/tests.rs (hunk 7 above)
Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler/tests.rs]: It actually turned out ok. I was expecting to make sky scrappers out of everything 🫢

File: rs/messaging/src/routing/stream_handler/tests.rs [hunk 8]
```diff
@@ -3468,281 +2870,599 @@ fn process_stream_slices_canister_migration_in_both_subnets_success() {
...
+#[allow(clippy::enum_variant_names)]
+#[derive(Copy, Clone)]
+enum MessageBuilder {
+    // `(sender, receiver)`.
+    BuildRequest(CanisterId, CanisterId),
+    // `(respondent, originator)`
+    BuildResponse(CanisterId, CanisterId),
+    // `(respondent, originator, reason)`
+    BuildRejectResponse(CanisterId, CanisterId, RejectReason),
+}
```
Comment (alin-at-dfinity) [inline rs/messaging/src/routing/stream_handler/tests.rs:3329]: Why not simply name the variants `Request`, `Response` (or `DataResponse`) and `RejectResponse`? I don't think we ever `use RequestRoResponse::*` or similar in this file, so there should be no conflict.

File: rs/messaging/src/routing/stream_handler/tests.rs (hunk 4 above)
Comment (alin-at-dfinity) [reply rs/messaging/src/routing/stream_handler/tests.rs]: I meant putting the test case name and documentation in terms of "a loopback stream with signals panics" instead of an implementation-specific and not very obvious "this index before that index panics".

I.e. something along the lines of:
```rust
/// Tests that inducting a loopback stream with signals panics: `induct_loopback_stream()`
/// always consumes all\ produced signals.
#[test]
#[should_panic(
    expected = "Invalid message indices in stream slice from subnet g24bn-xymaa-aaaaa-aaaap-yai: messages begin (21) != stream signals_end (22)"
)]
fn induct_loopback_stream_with_signals_panics() {
```

(Also, messages being GC-ed up to `sinals_end` and `signals_end` increasing monotonically doesn't really make any difference as to why we expect there to be no signals in the loopback stream.)

Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler/tests.rs]: My mental model was a bit different also considering the test just below this one. However I ditched the second part of your comment since without the knowledge that `induct_loopback_stream()` does produce signals but then immediately gc's them, this seems confusing.

File: rs/messaging/src/routing/stream_handler/tests.rs (hunk 8 above)
Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler/tests.rs]: I figured it might be confusing since I'm using `StreamConfig` with instructions on what kind of message we want in the setup stage and then also with actual messages in the 'construct the expected state' stage. Also there are a number of tests and functions that do use `Request` and `Response` but this can be resolved by using `ic_types::messages::Request` instead.

I guess it's not really possible to mistake `Request(*LOCAL_CANISTER, *LOCAL_CANISTER)` for an actual messages anyway so here you go.

File: rs/messaging/src/routing/stream_handler/tests.rs [hunk 9]
```diff
@@ -3468,281 +2870,599 @@ fn process_stream_slices_canister_migration_in_both_subnets_success() {
...
-        },
-        SignalConfig {
-            end: config.signals_end,
-            reject_signals: config.reject_signals,
-        },
-    )
+/// Returns a clone of the message in a stream at `stream_index`.
+///
+/// Panics if no such message exists.
+fn clone_message_from_stream(opt_stream: Option<&Stream>, stream_index: u64) -> RequestOrResponse {
```
Comment (alin-at-dfinity) [inline rs/messaging/src/routing/stream_handler/tests.rs:3251]: I find the `clone_message_from_[stream|slice]` functions unnecessary. Why not simply `[request|response|message]_in_[stream|slice]().clone()`? It would also make it unnecessary to have the `_ref_` in the method names above: just `request_in_stream()` / `response_in_slice()`; or even just `request_in()` for the former.

Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler/tests.rs:3251]: The `clone_message_from_()` functions are used where it doesn't matter which kind of message it is exactly. It also returns `RequestOrResponse` and not `&Request` or `&Response`. So what you are suggesting would be actually be `request_in_stream().clone().into()` which bloats the code for no reason IMO. It also puts emphasis on whether it's a request or a response in cases where it doesn't matter. I can dump the `ref` if you want.

I felt a small helper function like this doesn't matter much in a test and it expresses what I want without unnecessary details or `clone()`, `into()`, `unwrap()` and so forth. I am not entirely consistent with the `unwrap()` I guess you could say since I left it on inductions but not when getting messages from streams. I guess the former is far more likely to fail due to conceptual changes so `unwrap()` in place seemed like the better choice.

File: rs/messaging/src/routing/stream_handler/tests.rs [hunk 10]
```diff
@@ -83,520 +79,415 @@ fn oversized_reject_message_is_truncated() {
...
+                initial_available_guaranteed_response_memory;
+
+            let inducted_state = stream_handler
+                .induct_loopback_stream(state, &mut available_guaranteed_response_memory);
+
+            assert_eq!(expected_state, inducted_state);
+
+            // One reject response generated.
+            assert_eq!(
+                initial_available_guaranteed_response_memory - reject_response_num_bytes as i64,
```
[the remark this answers is not in the document]
Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler/tests.rs]: Renamed it.

File: rs/messaging/src/routing/stream_handler/tests.rs (hunk 9 above)
Comment (alin-at-dfinity) [reply rs/messaging/src/routing/stream_handler/tests.rs:3251]: I was more taking issue with the `clone_` prefix (and the function returning a clone). You could instead have `message_in_[stream|slice]()` functions (which would also allow you to drop the `_ref_` from other function names, since all of them return references now) and just write `message_in_stream(...).clone()` instead of `clone_message_in_stream(...)`.

A second concern (and the reason why I only had two helper functions, covering both streams and slices) is that the more magic you have in your test (`clone_message_from_stream()` vs `response_ref_in_slice()`, and so forth) the more you just have to rely on magic and hope you understand what it does. In (one sort of) an ideal world, tests would be written only using the public APIs of the component being tested, no access to the internals, no magic hiding what's going on.

I'm not insisting you go back to `get_request()` / `get_response()` (which I preferred), but I find a function with `clone_` in the name that simply clones whatever else it looks up, a bit unnecessary.

Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler/tests.rs:3251]: I guess I can make all these functions return references and then ditch the `clone`, `ref` in the name and put the `.clone()` behind the function. I still think not having to write `.into()` in lots of places merits this function.

File: rs/messaging/src/routing/stream_handler/tests.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (alin-at-dfinity) [reply rs/messaging/src/routing/stream_handler/tests.rs]: > I can clone it also but I don't see how that is more readable

To me, reading `assert_eq!(x + reject_response.count_bytes(), y)` is a lot more obvious than either `assert_eq!(x + reject_response_num_bytes, y)` or `assert_eq!(x + reject_response_count_bytes, y)`. No need to look through the code to see what that variable was initialized with (or just trust that I correctly guessed what it means). After all, tests are supposed to serve as documentation.

File: rs/messaging/src/routing/stream_handler/tests.rs (hunk 9 above)
Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler/tests.rs]: It's a double edged sword. Unfortunately `messages` is optional in `StreamSlice` which makes that annoying to deal with using the same function.

Overall it was quite cumbersome reading the original tests because there was always quite a bit of boilerplate in the way. I tried to find a good compromise of 'two much magic' and cutting off details obscuring what the test actually does. Especially in the later more complex tests where it was just generating identical messages for rerouted messages I felt it was a bit hard to follow what it going on. This is why I put all these `response @33` comments everywhere, though with the current framework it's not strictly necessary anymore.

I was also considering a sort of config to generate the expected stream but felt it was making things too stiff and not really any clearer. I guess I could give that it's own closure now, but I don't think it makes it better.

File: rs/messaging/src/routing/stream_handler/tests.rs [hunk 11]
```diff
@@ -83,520 +79,411 @@ fn oversized_reject_message_is_truncated() {
...
+    induct_loopback_stream_ignores_memory_limit_impl(HypervisorConfig {
+        subnet_message_memory_capacity: NumBytes::new(MAX_RESPONSE_COUNT_BYTES as u64 * 7 / 2),
+        subnet_wasm_custom_sections_memory_capacity: NumBytes::new(0),
+        ..Default::default()
     });
 }
 
+/// Common initial state setup for `StreamHandlerImpl::induct_loopback_stream()`
+/// memory limit tests.
+fn with_induct_loopback_stream_with_memory_limit_setup(
```
Comment (alin-at-dfinity) [inline rs/messaging/src/routing/stream_handler/tests.rs:466]: Nit: This is a very convoluted name. I'd at least drop the `_with_memory_limit` part, since that comes (or not) from the `HypervisorConfig` argument.

File: rs/messaging/src/routing/stream_handler/tests.rs [hunk 12]
```diff
@@ -689,1112 +552,1189 @@ fn induct_loopback_stream_with_memory_limit_impl(
...
+            stream_handler.garbage_collect_signals(
+                &mut streams.get_mut(&REMOTE_SUBNET).unwrap(),
+                REMOTE_SUBNET,
+                slices.get(&REMOTE_SUBNET).unwrap(),
+            );
+        },
+    );
 }
 
+/// Tests garbage collecting signals with a `StreamSlice` whose `message_begin` does not correspond
```
Comment (alin-at-dfinity) [inline rs/messaging/src/routing/stream_handler/tests.rs:814]: Nit:
```suggestion
/// Tests that garbage collecting signals with a `StreamSlice` whose `message_begin` does not correspond
```

Same everywhere applicable.

File: rs/messaging/src/routing/stream_handler/tests.rs [hunk 13]
```diff
@@ -689,1112 +552,1189 @@ fn induct_loopback_stream_with_memory_limit_impl(
...
+            reject_signals: vec![RejectSignal::new(RejectReason::QueueFull, 32.into())],
+            ..StreamSliceConfig::default()
+        }],
+        |stream_handler, state, slices, metrics| {
+            stream_handler.garbage_collect_local_state(state, &mut (i64::MAX / 2), &slices);
+
+            metrics.assert_eq_critical_errors(CriticalErrorCounts {
+                bad_reject_signal_for_response: 1,
+                ..CriticalErrorCounts::default()
+            });
```
Comment (alin-at-dfinity) [inline rs/messaging/src/routing/stream_handler/tests.rs:1208]: Nit: Not part of this MR, but since we're improving the tests, isn't it quite important part here to test what happened to the response?

File: rs/messaging/src/routing/stream_handler/tests.rs [hunk 14]
```diff
@@ -689,1112 +552,1189 @@ fn induct_loopback_stream_with_memory_limit_impl(
...
+        }],
+        |stream_handler, state, slices, metrics| {
+            // Mark `LOCAL_CANISTER` as having migrated from `CANISTER_MIGRATION_SUBNET` to `LOCAL_SUBNET`.
+            let state = simulate_canister_migration(
+                state,
+                *LOCAL_CANISTER,
+                CANISTER_MIGRATION_SUBNET,
+                LOCAL_SUBNET,
+            )
+            .unwrap();
```
Comment (alin-at-dfinity) [inline rs/messaging/src/routing/stream_handler/tests.rs:1508]: Do we ever not `unwrap()` the result of any of the `*_canister_migration()` functions? If not, I'd rather keep the `unwrap()` within the respective functions, they're merely test helpers / fixtures, no need to be pedantic about it.

File: rs/messaging/src/routing/stream_handler/tests.rs [hunk 15]
```diff
@@ -689,1112 +552,1189 @@ fn induct_loopback_stream_with_memory_limit_impl(
...
-            CANISTER_MIGRATION_SUBNET,
-            LOCAL_SUBNET,
-        );
+            messages: vec![
+                // ...a reject response for a request sent to `REMOTE_CANISTER` @0...
+                RejectResponse(*REMOTE_CANISTER, *LOCAL_CANISTER, RejectReason::QueueFull),
+                // ...a normal response @1...
+                Response(*REMOTE_CANISTER, *LOCAL_CANISTER),
+                // ...a request from @2.
+                Request(*REMOTE_CANISTER, *LOCAL_CANISTER),
```
Comment (alin-at-dfinity) [inline rs/messaging/src/routing/stream_handler/tests.rs:1496]:
```suggestion
                // ...a data response @1...
                Response(*REMOTE_CANISTER, *LOCAL_CANISTER),
                // ...and a request @2.
                Request(*REMOTE_CANISTER, *LOCAL_CANISTER),
```

File: rs/messaging/src/routing/stream_handler/tests.rs [hunk 16]
```diff
@@ -689,1112 +552,1189 @@ fn induct_loopback_stream_with_memory_limit_impl(
...
 ///
 /// `LOCAL_CANISTER` is marked as having migrated from `CANISTER_MIGRATION_SUBNET` to
 /// `LOCAL_SUBNET`.
 ///
-/// Tests that this results in the reject response accepted and inducted due the exception in
-/// `StreamHandler::should_accept_message_from()` and the normal response and request accepted
-/// and dropped.
+/// Tests that this results in the reject response accepted and inducted due to the exception in
+/// `StreamHandler::should_accept_message_from()` and the normal response and request are accepted
+/// but dropped.
```
Comment (alin-at-dfinity) [inline rs/messaging/src/routing/stream_handler/tests.rs:1478]: I must have written this test initially, but I can't for the life of me understand why we're dropping messages @1 and @2. Could you please clarify that in the doc comment?

File: rs/messaging/src/routing/stream_handler/tests.rs [hunk 17]
```diff
@@ -3468,281 +2859,603 @@ fn process_stream_slices_canister_migration_in_both_subnets_success() {
...
 
-fn get_response_at(queue: &StreamIndexedQueue<RequestOrResponse>, stream_index: u64) -> &Response {
-    match queue.get(stream_index.into()) {
-        Some(RequestOrResponse::Response(response)) => response,
-        _ => unreachable!(),
+/// Pushes the messages yielded by `iter` into the `state`.
+fn push_inputs<'a>(
+    state: &mut ReplicatedState,
+    iter: impl Iterator<Item = &'a RequestOrResponse>,
+) -> Result<(), (StateError, RequestOrResponse)> {
```
Comment (alin-at-dfinity) [inline rs/messaging/src/routing/stream_handler/tests.rs:3307]: I believe we always just `unwrap()` the output of `push_inputs()`, so we may as well do it here and not return anything.

I believe we do actually look at the output of `push_input()` once, but even in that case, I would rather have `push_input()` be infallible and either have a separate `try_push_input()` function or (better) simply have the one caller use `state.push_input()` directly.

File: rs/messaging/src/routing/stream_handler/tests.rs [hunk 18]
```diff
@@ -3248,28 +3239,23 @@ fn response_ref_in_slice(
 /// Returns a clone of the message in a stream at `stream_index`.
```
[the remark this answers is not in the document]
Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler/tests.rs]: Fixed.

File: rs/messaging/src/routing/stream_handler/tests.rs [hunk 19]
```diff
@@ -3248,28 +3239,23 @@ fn response_ref_in_slice(
...
-        Some(msg) => msg.clone(),
-        None => unreachable!(),
-    }
+fn message_in_stream(opt_stream: Option<&Stream>, stream_index: u64) -> &RequestOrResponse {
+    opt_stream
+        .and_then(|stream| stream.messages().get(stream_index.into()))
+        .unwrap()
 }
 
 /// Returns a clone of the message in a stream slice at `stream_index`.
```
[the remark this answers is not in the document]
Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler/tests.rs]: Fixed.

File: rs/messaging/src/routing/stream_handler/tests.rs (hunk 11 above)
Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler/tests.rs]: Renamed.

File: rs/messaging/src/routing/stream_handler/tests.rs (hunk 15 above)
Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler/tests.rs]: Fixed.

File: rs/messaging/src/routing/stream_handler/tests.rs (hunk 17 above)
Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler/tests.rs]: Removed the `unwrap()`. The one use where it used the error now just uses the naked state.

File: rs/messaging/src/routing/stream_handler/tests.rs (hunk 13 above)
Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler/tests.rs:1202]: True. The simplest way of checking this would be to refactor the test above to use the valid signal `CanisterMigrating` and then another time `EverythingElse` and check that the error counter is increment or not.

Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler/tests.rs]: Done. It's more precise and even a bit shorter.

File: rs/messaging/src/routing/stream_handler/tests.rs (hunk 16 above)
Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler/tests.rs]: Changed the doc comment as discussed offline.

File: rs/messaging/src/routing/stream_handler/tests.rs (hunk 12 above)
Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler/tests.rs]: Done.
